File: app/src/main/java/com/checkin/app/checkin/Waiter/WaiterRepository.java
package com.checkin.app.checkin.Waiter;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.BaseRepository;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.checkin.app.checkin.Waiter.Model.OrderStatusModel;
import com.checkin.app.checkin.Waiter.Model.SessionContactModel;
import com.checkin.app.checkin.Waiter.Model.WaiterEventModel;
import com.checkin.app.checkin.Waiter.Model.WaiterStatsModel;
import com.checkin.app.checkin.Waiter.Model.WaiterTableModel;
import com.checkin.app.checkin.misc.models.GenericDetailModel;
import com.checkin.app.checkin.session.models.CheckoutStatusModel;
import com.checkin.app.checkin.session.models.QRResultModel;
import com.checkin.app.checkin.session.models.RestaurantTableModel;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;

/**
 * Created by Shivansh Saini on 24/01/2019.
 */

public class WaiterRepository extends BaseRepository {
    private static WaiterRepository INSTANCE;
    private final WebApiService mWebService;

    private WaiterRepository(Context context) {
        mWebService = ApiClient.Companion.getApiService(context);
    }

    public static WaiterRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (WaiterRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new WaiterRepository(application.getApplicationContext());
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<List<WaiterTableModel>>> getWaiterServedTables() {
        return new NetworkBoundResource<List<WaiterTableModel>, List<WaiterTableModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<WaiterTableModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getWaiterServedTables());
            }

            @Override
            protected void saveCallResult(List<WaiterTableModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ObjectNode>> postSessionContact(long sessionId, SessionContactModel data) {
        return new NetworkBoundResource<ObjectNode, ObjectNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ObjectNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.postSessionContact(sessionId, data));
            }

            @Override
            protected void saveCallResult(ObjectNode data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<SessionContactModel>>> getSessionContacts(long sessionId) {
        return new NetworkBoundResource<List<SessionContactModel>, List<SessionContactModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<SessionContactModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getSessionContactList(sessionId));
            }

            @Override
            protected void saveCallResult(List<SessionContactModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<WaiterEventModel>>> getWaiterEventsForTable(long sessionId) {
        return new NetworkBoundResource<List<WaiterEventModel>, List<WaiterEventModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<WaiterEventModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getWaiterSessionEvents(sessionId));
            }

            @Override
            protected void saveCallResult(List<WaiterEventModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<RestaurantTableModel>>> getShopTables(long shopId,boolean activeQuery) {
        return new NetworkBoundResource<List<RestaurantTableModel>, List<RestaurantTableModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<RestaurantTableModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantTables(shopId,activeQuery));
            }

            @Override
            protected void saveCallResult(List<RestaurantTableModel> data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<QRResultModel>> newWaiterSession(ObjectNode requestJson) {
        return new NetworkBoundResource<QRResultModel, QRResultModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<QRResultModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postNewWaiterSession(requestJson));
            }

            @Override
            protected void saveCallResult(QRResultModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<OrderStatusModel>> changeOrderStatus(long orderId, ObjectNode data) {
        return new NetworkBoundResource<OrderStatusModel, OrderStatusModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<OrderStatusModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postChangeOrderStatus(orderId, data));
            }

            @Override
            protected void saveCallResult(OrderStatusModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<GenericDetailModel>> markEventDone(long eventId) {
        return new NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<GenericDetailModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.putSessionEventDone(eventId));
            }

            @Override
            protected void saveCallResult(GenericDetailModel data) {
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<WaiterStatsModel>> getWaiterStats(long restaurantId) {
        return new NetworkBoundResource<WaiterStatsModel, WaiterStatsModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<WaiterStatsModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantWaiterStats(restaurantId));
            }

            @Override
            protected void saveCallResult(WaiterStatsModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<CheckoutStatusModel>> postSessionRequestCheckout(long sessionId, ObjectNode data) {
        return new NetworkBoundResource<CheckoutStatusModel, CheckoutStatusModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<CheckoutStatusModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postSessionRequestCheckout(sessionId, data));
            }

            @Override
            protected void saveCallResult(CheckoutStatusModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<List<OrderStatusModel>>> postOrderListStatus(long sessionId, final List<OrderStatusModel> orders) {
        return new NetworkBoundResource<List<OrderStatusModel>, List<OrderStatusModel>>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<List<OrderStatusModel>>> createCall() {
                return new RetrofitLiveData<>(mWebService.postChangeOrderStatusList(sessionId, orders));
            }

            @Override
            protected void saveCallResult(List<OrderStatusModel> data) {

            }
        }.getAsLiveData();
    }
}

File: app/src/main/java/com/checkin/app/checkin/Waiter/WaiterRepository.kt
package com.checkin.app.checkin.Waiter

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.Waiter.Model.*
import com.checkin.app.checkin.data.BaseRepository
import com.checkin.app.checkin.data.db.dbStore
import com.checkin.app.checkin.data.network.ApiClient.Companion.getApiService
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.misc.models.GenericDetailModel
import com.checkin.app.checkin.session.models.*
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode
import io.objectbox.android.ObjectBoxLiveData

/**
 * Created by Shivansh Saini on 24/01/2019.
 */
class WaiterRepository private constructor(context: Context) : BaseRepository() {
    private val mWebService: WebApiService = getApiService(context)
    private val tableBox by dbStore<RestaurantTableModel>()
    private val tableSessionBox by dbStore<TableSessionModel>()

    val waiterServedTables: LiveData<Resource<List<WaiterTableModel>>>
        get() = object : NetworkBoundResource<List<WaiterTableModel>, List<WaiterTableModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<WaiterTableModel>>> {
                return RetrofitLiveData(mWebService.waiterServedTables)
            }

            override fun saveCallResult(data: List<WaiterTableModel>?) {}
        }.asLiveData

    fun postSessionContact(sessionId: Long, data: SessionContactModel): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.postSessionContact(sessionId, data))
            }

            override fun saveCallResult(data: ObjectNode?) {}
        }.asLiveData
    }

    fun getSessionContacts(sessionId: Long): LiveData<Resource<List<SessionContactModel>>> {
        return object : NetworkBoundResource<List<SessionContactModel>, List<SessionContactModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<SessionContactModel>>> {
                return RetrofitLiveData(mWebService.getSessionContactList(sessionId))
            }

            override fun saveCallResult(data: List<SessionContactModel>?) {}
        }.asLiveData
    }

    fun getWaiterEventsForTable(sessionId: Long): LiveData<Resource<List<WaiterEventModel>>> {
        return object : NetworkBoundResource<List<WaiterEventModel>, List<WaiterEventModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<WaiterEventModel>>> {
                return RetrofitLiveData(mWebService.getWaiterSessionEvents(sessionId))
            }

            override fun saveCallResult(data: List<WaiterEventModel>?) {}
        }.asLiveData
    }

    fun getShopTables(shopId: Long): LiveData<Resource<List<RestaurantTableModel>>> {
        return object : NetworkBoundResource<List<RestaurantTableModel>, List<RestaurantTableModel>>() {
            override fun shouldUseLocalDb(): Boolean = true

            override fun loadFromDb(): LiveData<List<RestaurantTableModel>>? {
                return ObjectBoxLiveData(tableBox.query()
                        .equal(RestaurantTableModel_.restaurantPk, shopId)
                        .build())
            }

            override fun createCall(): LiveData<ApiResponse<List<RestaurantTableModel>>> = RetrofitLiveData(mWebService.getRestaurantTables(shopId))

            override fun saveCallResult(data: List<RestaurantTableModel>?) = data?.run {
                forEach {
                    it.restaurantPk = shopId
                }
                tableBox.put(this)
            } ?: Unit
        }.asLiveData
    }

    fun filterShopTables(shopId: Long, activeQuery: Boolean): LiveData<Resource<List<RestaurantTableModel>>> {
        return object : NetworkBoundResource<List<RestaurantTableModel>, List<RestaurantTableModel>>() {
            override fun shouldUseLocalDb(): Boolean = true

            override fun loadFromDb(): LiveData<List<RestaurantTableModel>>? {
                return ObjectBoxLiveData(tableBox.query()
                        .equal(RestaurantTableModel_.restaurantPk, shopId)
                        .apply {
                            if (activeQuery) notNull(RestaurantTableModel_.relTableSessionId)
                            else isNull(RestaurantTableModel_.relTableSessionId)
                        }
                        .build())
            }

            override fun createCall(): LiveData<ApiResponse<List<RestaurantTableModel>>> = RetrofitLiveData(mWebService.filterRestaurantTables(shopId, activeQuery))

            override fun saveCallResult(data: List<RestaurantTableModel>?) = data?.run {
                forEach {
                    val savedTable = tableBox.get(it.qrPk)
                    if (savedTable != null) it.unseenEventCount = savedTable.unseenEventCount
                    it.restaurantPk = shopId
                }
                tableBox.put(this)
            } ?: Unit

            override fun transformResult(networkResult: Resource<List<RestaurantTableModel>>, dbResult: Resource<List<RestaurantTableModel>>?): Resource<List<RestaurantTableModel>> {
                return dbResult?.data?.let { dbList ->
                    if (networkResult.data == null) dbResult
                    else {
                        val result = networkResult.data.map { targetTable ->
                            val savedTable = dbList.find { it == targetTable }
                                    ?: return@map targetTable
                            targetTable.restaurantPk = savedTable.restaurantPk
                            targetTable.unseenEventCount = savedTable.unseenEventCount
                            targetTable
                        }
                        Resource.cloneResource(networkResult, result)
                    }
                } ?: networkResult
            }
        }.asLiveData
    }

    fun newWaiterSession(requestJson: ObjectNode): LiveData<Resource<QRResultModel>> {
        return object : NetworkBoundResource<QRResultModel, QRResultModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<QRResultModel>> {
                return RetrofitLiveData(mWebService.postNewWaiterSession(requestJson))
            }

            override fun saveCallResult(data: QRResultModel?) {}
        }.asLiveData
    }

    fun changeOrderStatus(orderId: Long, data: ObjectNode): LiveData<Resource<OrderStatusModel>> {
        return object : NetworkBoundResource<OrderStatusModel, OrderStatusModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<OrderStatusModel>> {
                return RetrofitLiveData(mWebService.postChangeOrderStatus(orderId, data))
            }

            override fun saveCallResult(data: OrderStatusModel?) {}
        }.asLiveData
    }

    fun markEventDone(eventId: Long): LiveData<Resource<GenericDetailModel>> {
        return object : NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<GenericDetailModel>> {
                return RetrofitLiveData(mWebService.putSessionEventDone(eventId))
            }

            override fun saveCallResult(data: GenericDetailModel?) {}
        }.asLiveData
    }

    fun getWaiterStats(restaurantId: Long): LiveData<Resource<WaiterStatsModel>> {
        return object : NetworkBoundResource<WaiterStatsModel, WaiterStatsModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<WaiterStatsModel>> {
                return RetrofitLiveData(mWebService.getRestaurantWaiterStats(restaurantId))
            }

            override fun saveCallResult(data: WaiterStatsModel?) {}
        }.asLiveData
    }

    fun postSessionRequestCheckout(sessionId: Long, data: ObjectNode): LiveData<Resource<CheckoutStatusModel>> {
        return object : NetworkBoundResource<CheckoutStatusModel, CheckoutStatusModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<CheckoutStatusModel>> {
                return RetrofitLiveData(mWebService.postSessionRequestCheckout(sessionId, data))
            }

            override fun saveCallResult(data: CheckoutStatusModel?) {}
        }.asLiveData
    }

    fun postOrderListStatus(sessionId: Long, orders: List<OrderStatusModel>): LiveData<Resource<List<OrderStatusModel>>> {
        return object : NetworkBoundResource<List<OrderStatusModel>, List<OrderStatusModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<OrderStatusModel>>> {
                return RetrofitLiveData(mWebService.postChangeOrderStatusList(sessionId, orders))
            }

            override fun saveCallResult(data: List<OrderStatusModel>?) {}
        }.asLiveData
    }

    companion object : SingletonHolder<WaiterRepository, Application>({ WaiterRepository(it.applicationContext) })
}